apps/sparo-lib/src/services/SparoProfileService.ts

```diff
@@ -99,4 +108,192 @@ export class SparoProfileService {
+  }: {
+    profiles?: Set<string>;
+    addProfiles?: Set<string>;
+  }): Promise<void> {
+    /*
+     * 2. If profile array is specified, using `git sparse-checkout set` to set sparse checkout folders in profiles.
+     * 3. If add profiles is specified, using `git sparse-checkout add` to add folders in add profiles
+     */
+
+    this._localState.reset();
+    this._terminalService.terminal.writeLine(
+      `Syncing local sparse checkout state with following specified profiles:\n${Array.from([
+        ...(profiles ?? []),
+        ...(addProfiles ?? [])
+      ]).join('\n')}`
+    );
+    this._terminalService.terminal.writeLine();
+    if (!profiles || profiles.size === 0) {
+      // If no profile was specified, purge local state to skeleton
+      await this._gitSparseCheckoutService.purgeAsync();
+    } else {
+      const allProfilesIncludeFolders: string[] = [],
+        allProfilesExcludeFolders: string[] = [],
+        allProfilesSelections: ISelection[] = [];
+      for (const profile of profiles) {
+        // Since we have run localState.reset() before, for each profile we just add it to local state.
+        const { selections, includeFolders, excludeFolders } = await this.resolveSparoProfileAsync(profile, {
+          localStateUpdateAction: 'add'
+        });
+        // combine all profiles' selections and include/exclude folder
+        allProfilesSelections.push(...selections);
+        allProfilesIncludeFolders.push(...includeFolders);
+        allProfilesExcludeFolders.push(...excludeFolders);
+      }
+      // sparse-checkout set once for all profiles together
+      await this._gitSparseCheckoutService.checkoutAsync({
+        selections: allProfilesSelections,
+        includeFolders: allProfilesIncludeFolders,
+        excludeFolders: allProfilesExcludeFolders,
+        checkoutAction: 'set'
+      });
+    }
+    if (addProfiles?.size) {
+      // If add profiles is specified, using `git sparse-checkout add` to add folders in add profiles
+      const allAddProfilesSelections: ISelection[] = [],
+        allAddProfilesIncludeFolders: string[] = [],
+        allAddProfilesExcludeFolders: string[] = [];
+      for (const profile of addProfiles) {
+        // For each add profile we add it to local state.
+        const { selections, includeFolders, excludeFolders } = await this.resolveSparoProfileAsync(profile, {
+          localStateUpdateAction: 'add'
+        });
+        // combine all add profiles' selections and include/exclude folder
+        allAddProfilesSelections.push(...selections);
+        allAddProfilesIncludeFolders.push(...includeFolders);
+        allAddProfilesExcludeFolders.push(...excludeFolders);
+      }
+      /**
+       * Note:
+       * Although we could run sparse-checkout add multiple times,
+       * we combine all add operations and execute once for better performance.
+       */
+      await this._gitSparseCheckoutService.checkoutAsync({
+        selections: allAddProfilesSelections,
```

Comment:
Related to [above comment](https://github.com/tiktok/sparo/pull/28/files/b3771c306cba2295daa0c58bfc687ba7db6be249#r1507969225), I'm still wondering whether it might be more efficient to generate `.git/info/sparse-checkout` file directly, rather than invoking `git sparse-checkout`:

https://github.com/tiktok/sparo/blob/dc297429534cc894ce243a103e37c30733c6a182/apps/sparo-lib/src/services/GitSparseCheckoutService.ts#L258-L264

Besides avoiding the overhead of spawning a process with a very long command line, perhaps the globs themselves could be represented more efficiently than the somewhat odd syntax generated by `git sparse-checkout`: (?)

```
/*
!/*/
/apps/
!/apps/*/
/apps/my-app/
!/apps/my-app/*/
/apps/my-app/_/
```

(It would be implemented in a future PR of course.)